### PR TITLE
Add F-score metric to metrics.py

### PR DESCRIPTION
**Pull request by carlthome**
# Proposal
For multi-label classification it is useful to not only measure how many samples were labeled with the right class, but also how many that were incorrect, otherwise a model would achieve perfect accuracy by assigning every class to every sample.

A way of measuring multi-label performance is to take both precision and recall into account, and the most common way of combining these into a single scalar value is by calculating a [F-measure](https://en.wikipedia.org/wiki/F1_score). This pull request adds a F-measure function to metrics.py.

For reference, [sklearn has this already](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html).

# Recommended reading
>Jansche, Martin. "Maximum expected F-measure training of logistic regression models." Proceedings of the conference on Human Language Technology and Empirical Methods in Natural Language Processing. Association for Computational Linguistics, 2005.

### CODE REVIEW

**Comment by kilotaras** [comment]
`count_nonzero(y_pred)` would usually be close to total number of elements, as you generally get numbers close to 0, but not exactly.

**Comment by carlthome** [comment]
@yxonic, thanks! I'll fix that. 👍 

@kilotaras, would you recommend having a small threshold instead for when a value is to be considered a positive/negative sample?

**Comment by carlthome** [comment]
@yxonic and @kilotaras, how about this?

**Comment by carlthome** [comment]
In order to be consistent with previous metrics I'm going with the convention that if the result of `K.round` is zero the sample was negative, and if the rounding is one the sample was positive. This seems fair enough as sigmoids are centered around 0.5 and are probably the most typical activation function for multi-label classification.

**Comment by carlthome** [comment]
@fchollet, thoughts?

File: keras/metrics.py [hunk 1]
```diff
@@ -91,6 +91,32 @@ def matthews_correlation(y_true, y_pred):
 
     return numerator / (denominator + K.epsilon())
 
+
+def fmeasure(y_true, y_pred, beta=1):
```
Comment (fchollet) [inline keras/metrics.py]: Should we name it `f1_score` for consistency with sklearn?

File: keras/metrics.py [hunk 2]
```diff
@@ -91,6 +91,32 @@ def matthews_correlation(y_true, y_pred):
...
+    '''Approximate F-score as an overall loss for all the samples at once.
```
Comment (fchollet) [inline keras/metrics.py]: This docstring will become part of the documentation, so it will need to be more informative. See the sklearn docs for the same feature for inspiration.

File: keras/metrics.py (hunk 1 above)
Comment (carlthome) [inline keras/metrics.py]: No, because `beta` is a parameter and not fixed to one. [fbeta_score](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.fbeta_score.html) has precedent. Would that work?

File: keras/metrics.py (hunk 2 above)
Comment (carlthome) [inline keras/metrics.py]: Will do!

File: keras/metrics.py (hunk 1 above)
Comment (carlthome) [inline keras/metrics.py]: Fixed.

File: keras/metrics.py (hunk 2 above)
Comment (carlthome) [inline keras/metrics.py]: Fixed.

**Comment by fchollet** [comment]
PEP8 test is failing: https://travis-ci.org/fchollet/keras/jobs/165277680

Please use a PEP8 linter to avoid such issues.

**Comment by carlthome** [comment]
@fchollet, green now. Merge?

**Comment by cgratie** [comment]
Having F-score (and AUC scores) available in Keras would be really useful. On the other hand I feel that the current implementation has a major drawback because it ignores the way Keras aggregates metrics internally. What I mean is that per-epoch F-score will only be correct if there is a single batch in the epoch (and the batch is large enough).

Here's why this is the case: most metrics available in Keras are averages of some metric that can be computed for each sample, and they have K.mean as the top operation in the corresponding tensor. When aggregating the metrics of several batches together, Keras internally multiplies the per-batch value with the number of samples in the batch and keeps a running total. At the end of the epoch, the running total is divided by the total number of samples seen, and this gives the per-epoch value of the metric.

On the other hand, this approach is not compatible with metrics that are not averages of a per-sample metric. In other words, the reported per-epoch F-measure will simply be the mean of the per-batch F-measures, **which is not the same as the actual F-measure of the epoch**.

Any thoughts on how Keras could accommodate such metrics correctly?

**Comment by fchollet** [comment]
> Any thoughts on how Keras could accommodate such metrics correctly?

It is trivial to write such metrics by using a callback.

**Comment by carlthome** [comment]
@cgratie, I think the progress reporting of the metrics has been through a bit of a balancing act between feeling responsive enough and being precise enough, hence the batchwise aggregating. If an exact evaluation is required for every epoch (instead of just at test time), people tend to explicitly call `model.evaluate` between epochs (or better yet, use callbacks like @fchollet suggests). As far as I know it's impossible to calculate the F-score during training, and it will always be an approximation.

**Comment by cgratie** [comment]
@fchollet Yes, a callback is what I am using at the moment, but I do like the metrics interface better.
@carlthome I hadn't considered the impact of dropout and similar things, good point, it would just be an approximation anyway. 

On the other hand I feel that it could be possible to compute this approximation accurately during training, only one would need to aggregate false positives, true positives etc and use those to compute the f-score (and other metrics that rely on them). I could attempt to implement this if you guys think it would be useful.

**Comment by carlthome** [comment]
I think that could be cool, but it's not straightforward how such a function could fit into metrics.py without major refactoring, aside from using a stateful closure which I think would be frowned upon.

**Comment by cgratie** [comment]
Definitely not straightforward, indeed. What I'm thinking is something similar to how the base logger does it now, but it would require the use of some "hidden" metrics to compute the values that can be aggregated over an epoch (true positives, false positives, etc). I'll come back with a pull request if I ever manage to write this with not too much refactoring.
